source/slang/ir.h

```diff
@@ -991,23 +981,50 @@ struct IRModuleInst : IRParentInst
 
 struct IRModule : RefObject
 {
-    // The compilation session in use.
-    Session*    session;
-    MemoryPool  memoryPool;
-    List<IRObject*> irObjectsToFree; // list of ir objects to run destructor upon destruction
+    enum 
+    {
+        kMemoryArenaBlockSize = 16 * 1024,           ///< Use 16k block size for memory arena
+    };
 
-    IRModuleInst* moduleInst;
-    IRModuleInst* getModuleInst() { return moduleInst;  }
+    SLANG_FORCE_INLINE Session* getSession() const { return session; }
+    SLANG_FORCE_INLINE IRModuleInst* getModuleInst() const { return moduleInst;  }
 
-    IRInstListBase getGlobalInsts() { return getModuleInst()->getChildren(); }
+    IRInstListBase getGlobalInsts() const { return getModuleInst()->getChildren(); }
 
-    ~IRModule()
+    void addRefObjectToFree(RefObject* obj)
+    {
+        if (obj)
+        {
+            obj->addReference();
+            m_refObjectsToFree.Add(obj);
+        }
+    }
+    
+        /// Ctor
+    IRModule():
+        memoryArena(kMemoryArenaBlockSize)
     {
-        for (auto val : irObjectsToFree)
-            if (!val->isDestroyed)
-                val->dispose();
-        irObjectsToFree = List<IRObject*>();
     }
+
+    ~IRModule()
+    { 
+        // Release all ref objects
+        for (RefObject* ptr: m_refObjectsToFree)
+        {
+            ptr->releaseReference();
+        }
+        // Clear any memory too
+        m_refObjectsToFree = List<RefObject*>();
+    }
+
+    MemoryArena memoryArena;
```

Comment:
What's the rationale for moving around the data members, including the ones that weren't affected by the change? I understand that there are some common coding conventions in which people emphasize putting the methods first, but most of those insist that all the data should be `private` or `protected`.

In the case of Slang, we don't currently put a high value on hiding stuff with `private` or `protected`, except when a class has very subtle invariants that are easy to mess up by accident (or its a fundamental type where enforcing those invariants will pay off across many uses). When working in codebases that are much more zealous about things like `private` and `const` I tend to find that the friction of dealing with those mechanisms outweighs the (sometimes purely theoretical) benefits they provide.

Putting all the data first lets us focus on the in-memory representation of things, and makes it easier to micromanage layout and/or the order things will be listed in debuggers by default.

Also: I'm still not 100% sold on the `m_` prefix as a matter of policy. :)